### PR TITLE
Dynamic options not handled in PDF

### PR DESCRIPTION
**Pull request by RonnyB71**
This fixes the issue where only the value of an dynamic option based field was displayed in the PDF.

### CODE REVIEW

**Review by lorang92 — COMMENTED** [review]
At a quick glance it still seems to me that the `GetOptionIdsFromFormLayout` returns one key for each unique option id. (If this is due to my lack of understanding in c# please correct me).

What happens if a layout is setup with two dropdowns that is referencing the same optionId but with different params? 

For instance: 

```json
    "layout": [
      {
        "id": "dropdown-component-1",
        "type": "Dropdown",
        "textResourceBindings": {
          "title": "Some title",
          "description": "Some description"
        },
        "dataModelBindings": {
          "simpleBinding": "Felt1"
        },
        "required": true,
        "optionsId": "test",
        "mapping": {
          "some.field": "someArg"
        }
      },
      {
        "id": "dropdown-component-2",
        "type": "Dropdown",
        "textResourceBindings": {
          "title": "Some title",
          "description": "Some description"
        },
        "dataModelBindings": {
          "simpleBinding": "Felt2"
        },
        "required": true,
        "optionsId": "test",
        "mapping": {
          "some.other.field": "someArg"
        }
      }
    ]
```
I think we have to support this? In frontend we create a key using the values in the mapping, thus making it unique for the same option id but with different mapping values.

**Comment by RonnyB71** [comment]
> At a quick glance it still seems to me that the `GetOptionIdsFromFormLayout` returns one key for each unique option id. (If this is due to my lack of understanding in c# please correct me).
> 
> What happens if a layout is setup with two dropdowns that is referencing the same optionId but with different params?
> 
> For instance:
> 
> ```json
>     "layout": [
>       {
>         "id": "dropdown-component-1",
>         "type": "Dropdown",
>         "textResourceBindings": {
>           "title": "Some title",
>           "description": "Some description"
>         },
>         "dataModelBindings": {
>           "simpleBinding": "Felt1"
>         },
>         "required": true,
>         "optionsId": "test",
>         "mapping": {
>           "some.field": "someArg"
>         }
>       },
>       {
>         "id": "dropdown-component-2",
>         "type": "Dropdown",
>         "textResourceBindings": {
>           "title": "Some title",
>           "description": "Some description"
>         },
>         "dataModelBindings": {
>           "simpleBinding": "Felt2"
>         },
>         "required": true,
>         "optionsId": "test",
>         "mapping": {
>           "some.other.field": "someArg"
>         }
>       }
>     ]
> ```
> 
> I think we have to support this? In frontend we create a key using the id + stringify the mapping, thus making it unique for the same option id but with different mapping values.

Nice catch and good point. I'll rewrite to support this and add a new test as well.

File: src/Altinn.App.PlatformServices/Implementation/AppResourcesSI.cs [hunk 1]
```diff
@@ -437,5 +431,24 @@ private byte[] ReadFileContentsFromLegalPath(string legalPath, string filePath)
 
             return null;
         }
+        
+        /// <inheritdoc />
+        [Obsolete("GetOptions method is obsolete and will be removed in the future. Use the corresponding method in IAppOptions interface instead.", false)]
```
Comment (tjololo) [inline src/Altinn.App.PlatformServices/Implementation/AppResourcesSI.cs:436]: If we are planning a new major version with breaking changes. Should we consider changing the error param so it generates compile error instead of a warning?

File: src/Altinn.App.PlatformServices/Implementation/AppResourcesSI.cs [hunk 2]
```diff
@@ -437,5 +431,24 @@ private byte[] ReadFileContentsFromLegalPath(string legalPath, string filePath)
...
+        public List<AppOption> GetOptions(string optionId)
+        {
+            return _appOptionsService.GetOptions(optionId);
+        }
+
+        /// <inheritdoc/>
+        public async Task<AppOptions> GetOptionsAsync(string optionId, string language, Dictionary<string, string> keyValuePairs)
```
Comment (tjololo) [inline src/Altinn.App.PlatformServices/Implementation/AppResourcesSI.cs:443]: Are these also Obsolete since `GetOptions(string)`is Obsolete? Should they even be implemented?

File: src/Altinn.App.PlatformServices/Implementation/AppBase.cs [hunk 3]
```diff
@@ -43,17 +35,13 @@ public abstract class AppBase : IAltinnApp
         private readonly IAppResources _resourceService;
         private readonly IProcess _processService;
```
Comment (tjololo) [inline src/Altinn.App.PlatformServices/Implementation/AppBase.cs:36]: If this is really unused should we remove it?

**Review by tjololo — COMMENTED** [review]
Still getting to know the code base so can't review all of the logic

File: src/Altinn.App.PlatformServices/Implementation/AppResourcesSI.cs (hunk 1 above)
Comment (RonnyB71) [reply src/Altinn.App.PlatformServices/Implementation/AppResourcesSI.cs:436]: Good point, went one step further and removed the method since we are creating a new major version.

File: src/Altinn.App.PlatformServices/Implementation/AppResourcesSI.cs (hunk 2 above)
Comment (RonnyB71) [reply src/Altinn.App.PlatformServices/Implementation/AppResourcesSI.cs:443]: No they are not obsolete as such, but they are used here as a passthrough as one step on the way to split out options related methods from IAppResources into it's own interface (IAppOptionsService). The IAppResources is to big and should be split into smaller more targeted interfaces. Established the IAppOptionsService but kept the methods in IAppResources to avoid too much refactoring and breaking changes the last time, but since we now have a breaking change and it's only PDF left that uses the methods from AppResourcesSI/IAppResources they can be switched out with a dependency to IAppOptionsService instead. 

Long story short, I'm removing them as you suggest:)

File: src/Altinn.App.PlatformServices/Implementation/AppBase.cs (hunk 3 above)
Comment (RonnyB71) [reply src/Altinn.App.PlatformServices/Implementation/AppBase.cs:36]: Fortgot about these, yes I'll remove them.

File: src/Altinn.App.PlatformServices/Implementation/PdfService.cs [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            IEnumerable<JToken> componentsWithOptionsDefined = GetFormComponentsWithOptionsDefined(formLayout);
+
+            Dictionary<string, Dictionary<string, string>> dictionary = new Dictionary<string, Dictionary<string, string>>();
+
+            foreach (JToken component in componentsWithOptionsDefined)
+            {
+                string optionsId = component.SelectToken("optionsId").Value<string>();
+                bool hasMappings = component.SelectToken("mapping") != null;
+                Dictionary<string, string> keyValuePairs = hasMappings ? GetComponentKeyValuePairs(component, data) : new Dictionary<string, string>();
+                AppOptions appOptions = await _appOptionsService.GetOptionsAsync(optionsId, language, keyValuePairs);
```
Comment (TheTechArch) [inline src/Altinn.App.PlatformServices/Implementation/PdfService.cs:208]: Kunne det være aktuelt å trekke dette utenfor løkken. Hvis de er samme options, eller satser vi bare på caching i service.

File: src/Altinn.App.PlatformServices/Implementation/PdfService.cs [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+            foreach (var map in mappings)
+            {
+                JToken selectedData = jsonData.SelectToken(map.Key);
+                componentKeyValuePairs.Add(map.Value, selectedData.ToString());
+            }
+
+            return componentKeyValuePairs;
+        }
+        
+        private static Dictionary<string, string> GetMappingsForComponent(JToken component)
```
Comment (TheTechArch) [inline src/Altinn.App.PlatformServices/Implementation/PdfService.cs:244]: Støtter vi options i repeterende gruppe med denne magien?

**Review by TheTechArch — APPROVED** [review]
Looks. Good. Not able to verify all changes in test apps, but trust the compiler

File: src/Altinn.App.PlatformServices/Implementation/PdfService.cs (hunk 4 above)
Comment (RonnyB71) [reply src/Altinn.App.PlatformServices/Implementation/PdfService.cs:208]: Hvis det er flere felter som er koblet til samme options, så tenker jeg det er mest naturlig at OptionsService'en tar logikken med å cache. Den må da se på kombinasjonen av optionsId, language og keyValuePairs. Foreløpig har vi ikke dette, men det burde vi nok få på plass.

File: src/Altinn.App.PlatformServices/Implementation/PdfService.cs (hunk 5 above)
Comment (RonnyB71) [reply src/Altinn.App.PlatformServices/Implementation/PdfService.cs:244]: Nei. Den logikken ligger i frontend og er per i dag duplisert inn i PdfService'en. Men det er et godt poeng at når vi nå skal gjøre noe større med PdfServicen så bør vi i størst mulig grad samle logikken.
